### PR TITLE
Add jacobian-vector product support

### PR DESCRIPTION
**Pull request by spcornelius**
These changes add support for automatically derived jacobian-vector product functions within
SymbolicSys. This can be used for integration with CVODES, either through pycvodes or native code.

### CODE REVIEW

**Comment by spcornelius** [comment]
I was not quite sure the best approach to some of the test failures, which stem (in part) from some compatibility-breaking things I introduced into pycvodes that have not been released yet. (Although there do seem to be a few pre-existing errors, notably "'NativeCvodeCode' object has no attribute '_remove_tempdir_on_clean'"--which appears to go away with unreleased pycodeexport content?--and an unused variable error that I should have fixed yesterday  in https://github.com/bjodah/pycvodes/pull/76.

Anyway, the new compilation errors come from my 1) renaming jac_times_vec ---> jtimes and 2) taking in linear_solver as a str, not int in pycvodes. There seem to be a couple unsavory options to deal with this:

1) Keep pyodesys fully compatibility with the latest release of pycvodes, updating the interface of the former only after a new release of the latter.
2) Match the interface here to the current development version of pycvodes; have the CI tests fetch that version instead of installing pycvodes
3) Some hybrid interface for pyodesys that works with both old and development. Have not thought about how to do this elegantly...

I obviously defer to you on this.

**Comment by bjodah** [comment]
Let's do 1) or 2). I could mint new releases tomorrow if that sounds alright by you?

**Comment by bjodah** [comment]
So I've started updating dependent packages. (Not finished quite yet).
Since the includes of headers from AnyODE will depend on a macro (whether to use blasless version or not) I think I'm leaning towards renaming that macro (`USE_LAPACK`) to (`ANYODE_NO_LAPACK`) and have users actively opting out of lapack instead of the opposite (it will be less disruptive in some of my private packages as well). Does that sound alright by you?

I'll get back to updating the interfaces tomorrow and will hopefully be able to upload new releases of all the dependencies then as well.

**Comment by bjodah** [comment]
I added a few commits here (made AnyODE a submodule, updated CI config with suitesparse).
When I run tests locally now the first error reported looks like it's from the new jtimes facilities:
```
$ CPATH=/opt/sundials-3.1.1/include:/usr/include/suitesparse LIBRARY_PATH=/opt/sundials-3.1.1/lib LD_LIBRARY_PATH=/opt/sundials-3.1.1/lib ./scripts/run_tests.sh -x
========================================== test session starts ==========================================
platform linux -- Python 3.5.2, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /home/bjorn/vc/pyodesys, inifile: setup.cfg
plugins: pudb-0.6, pep8-1.0.6, flakes-2.0.0
collected 349 items                                                                                     

conftest.py s.                                                                                    [  0%]
setup.py ..                                                                                       [  0%]
examples/_chem_kinet.py s.                                                                        [  0%]
examples/pydy_double_pendulum.py s.                                                               [  1%]
examples/van_der_pol.py s.                                                                        [  1%]
examples/tests/test_examples.py s....                                                             [  2%]
pyodesys/__init__.py s.                                                                           [  2%]
pyodesys/_release.py s.                                                                           [  3%]
pyodesys/analysis.py s.                                                                           [  3%]
pyodesys/convergence.py s.                                                                        [  3%]
pyodesys/core.py F

=============================================== FAILURES ================================================
____________________________________________ pyflakes-check _____________________________________________
/home/bjorn/vc/pyodesys/pyodesys/core.py:683: UndefinedName
undefined name '_p'
/home/bjorn/vc/pyodesys/pyodesys/core.py:684: UndefinedName
undefined name '_p'
```
So I'll hand this over back to you from here. Let me know if I can help.

**Comment by bjodah** [comment]
I released pyodesys-0.11.15 with new (upper) limits on dependencies:
https://github.com/bjodah/pyodesys/pull/85

You'll need to invert those for your branch (which we'll release as pyodesys-0.12.0).

**Comment by spcornelius** [comment]
Finally getting back to this now that the  (I think) all the requisite changes  have been pushed to anyode/pycvodes/sym. Given the substantial changes in those, I don't know how you want to proceed. Let me reply to your other remarks here first.

File: pyodesys/core.py [hunk 1]
```diff
@@ -667,14 +667,26 @@ def _integrate_cvode(self, *args, **kwargs):
...
             raise ValueError("lband and uband set locally (set at"
                              " initialization instead)")
         if self.band is not None:
             kwargs['lband'], kwargs['uband'] = self.band
         kwargs['autonomous_exprs'] = self.autonomous_exprs
+
+        with_jtimes = kwargs.pop('with_jtimes', False)
+
+        if with_jtimes is True:
+            def _jtimes(v, Jv, x, y, fy=None, user_data=None, tmp=None):
```
[the remark this answers is not in the document]
Comment (spcornelius) [reply pyodesys/core.py]: user_data & tmp were dropped from AnyODE (including in PyODESys). So I suppose it can be dropped here.

**Comment by bjodah** [comment]
Feel free to replace it with a submodule. Breaking backwards compatibility is also fine (as long as we retain capabilities). I'll maintain the pyodesys-0.11 branch as a legacy version (to give myself and others time to adapt to 0.12 once you feel it's time to tag it).

**Pull request closed without merging**